// @uxpin/merge-cli/src/program/command/experimentation/steps/experimentationBuildLibraryStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuildOptions = exports.experimentationBuildLibraryStep = void 0;
const getProjectRoot_1 = require("../../../args/providers/paths/getProjectRoot");
const getTempDirPath_1 = require("../../../args/providers/paths/getTempDirPath");
const thunkBuildComponentsLibrary_1 = require("../../../utils/thunkBuildComponentsLibrary");
function experimentationBuildLibraryStep(args) {
    return { exec: thunkBuildComponentsLibrary_1.thunkBuildComponentsLibrary(getBuildOptions(args)), shouldRun: true };
}
exports.experimentationBuildLibraryStep = experimentationBuildLibraryStep;
function getBuildOptions(args) {
    const { webpackConfig, wrapper } = args;
    return {
        development: true,
        projectRoot: getProjectRoot_1.getProjectRoot(args),
        uxpinDirPath: getTempDirPath_1.getTempDirPath(args),
        webpackConfigPath: webpackConfig,
        wrapperPath: wrapper,
    };
}
exports.getBuildOptions = getBuildOptions;
