// @uxpin/merge-cli/src/steps/serialization/component/presets/jsx/bundle/__tests__/getSourceFileContentToBundle.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const getSourceFileContentToBundle_1 = require("../getSourceFileContentToBundle");
describe('getSourceFileContentToBundle', () => {
    it('should get file content for given presets', () => {
        // given
        const components = [
            {
                defaultExported: true,
                documentation: { examples: [] },
                info: {
                    dirPath: '',
                    implementation: { framework: 'reactjs', lang: 'javascript', path: '' },
                    presets: [
                        { path: './src/components/Component1/presets/0-default.jsx' },
                        { path: './src/components/Component1/presets/1-alternative.jsx' },
                    ],
                },
                name: 'Component1',
                presets: [],
                properties: [],
            },
            {
                defaultExported: true,
                documentation: { examples: [] },
                info: {
                    dirPath: '',
                    implementation: { framework: 'reactjs', lang: 'javascript', path: '' },
                    presets: [
                        { path: './src/components/Component2/presets/0-default.jsx' },
                    ],
                },
                name: 'Component2',
                presets: [],
                properties: [],
            },
        ];
        // when
        const result = getSourceFileContentToBundle_1.getSourceFileContentToBundle('./', components);
        // then
        expect(result).toMatchSnapshot();
    });
    it('should get file content for empty presets list', () => {
        // given
        const components = [];
        // when
        const result = getSourceFileContentToBundle_1.getSourceFileContentToBundle('./', components);
        // then
        expect(result).toMatchSnapshot();
    });
});
