// @uxpin/merge-cli/src/steps/experimentation/server/handler/file/StaticFileHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticFileHandler = void 0;
const tslib_1 = require("tslib");
const http_status_codes_1 = require("http-status-codes");
const readFileFromPath_1 = require("../../../../../utils/fs/readFileFromPath");
const decorateWithAccessControlAllowOrigin_1 = require("../../headers/decorateWithAccessControlAllowOrigin");
const NotFoundHandler_1 = require("../error/NotFoundHandler");
class StaticFileHandler {
    constructor(filePath, headers = {}) {
        this.filePath = filePath;
        this.headers = headers;
        this.notFoundHandler = new NotFoundHandler_1.NotFoundHandler();
    }
    handle(request, response) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            response.writeHead(http_status_codes_1.OK, decorateWithAccessControlAllowOrigin_1.decorateWithAccessControlAllowOrigin(this.headers, request.headers));
            try {
                const content = yield this.readFileContent();
                response.end(content, 'utf-8');
            }
            catch (error) {
                this.notFoundHandler.handle(request, response);
            }
        });
    }
    readFileContent() {
        return readFileFromPath_1.readFileFromPath(this.filePath);
    }
}
exports.StaticFileHandler = StaticFileHandler;
