// @uxpin/merge-cli/src/steps/serialization/component/props/parsing/descriptors/parseTypeTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTypeTag = void 0;
const ComponentPropertyDefinition_1 = require("../../../implementation/ComponentPropertyDefinition");
function parseTypeTag(value) {
    if (!value) {
        return;
    }
    const typeMatch = value.match(/[a-z]+/);
    if (!typeMatch) {
        return;
    }
    switch (typeMatch[0]) {
        case ComponentPropertyDefinition_1.CustomControlTypeName.CodeEditor:
        case ComponentPropertyDefinition_1.CustomControlTypeName.Input:
        case ComponentPropertyDefinition_1.CustomControlTypeName.Interactions:
        case ComponentPropertyDefinition_1.CustomControlTypeName.Number:
        case ComponentPropertyDefinition_1.CustomControlTypeName.Select:
        case ComponentPropertyDefinition_1.CustomControlTypeName.Switcher: {
            return parseCustomType(typeMatch[0]);
        }
        case ComponentPropertyDefinition_1.CustomControlTypeName.Textfield: {
            return parseTextfieldType(value);
        }
        default: {
            return;
        }
    }
}
exports.parseTypeTag = parseTypeTag;
const TEXTFIELD_REGEX = /(^textfield$|^textfield(\(([\d]+)?\)))/;
const TEXTFIELD_DEFAULT_ROWS = 3;
const ROWS_MATCH_ID = 3;
function parseTextfieldType(value) {
    const match = value.match(TEXTFIELD_REGEX);
    if (!match) {
        return;
    }
    const rows = isNaN(parseInt(match[ROWS_MATCH_ID], 10))
        ? TEXTFIELD_DEFAULT_ROWS
        : parseInt(match[ROWS_MATCH_ID], 10);
    return {
        serialized: {
            customType: {
                name: ComponentPropertyDefinition_1.CustomControlTypeName.Textfield,
                structure: {
                    rows: Math.max(rows, 1),
                },
            },
        },
        type: ComponentPropertyDefinition_1.CustomDescriptorsTags.TYPE,
    };
}
function parseCustomType(typeName) {
    return {
        serialized: {
            customType: {
                name: typeName,
                structure: {},
            },
        },
        type: ComponentPropertyDefinition_1.CustomDescriptorsTags.TYPE,
    };
}
