// @uxpin/merge-cli/src/steps/experimentation/server/handler/code/GetCategoriesHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCategoriesHandler = exports.COMPONENT_TYPE = void 0;
const tslib_1 = require("tslib");
const http_status_codes_1 = require("http-status-codes");
const getProjectMetadata_1 = require("../../../metadata/getProjectMetadata");
const getAccessControlHeaders_1 = require("../../headers/getAccessControlHeaders");
const getNoCacheHeaders_1 = require("../../headers/getNoCacheHeaders");
exports.COMPONENT_TYPE = 'code-sync-component';
class GetCategoriesHandler {
    constructor(context) {
        this.context = context;
    }
    handle(request, response) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            response.writeHead(http_status_codes_1.OK, Object.assign(Object.assign({ 'Content-Type': 'application/json' }, getAccessControlHeaders_1.getAccessControlHeaders(request.headers)), getNoCacheHeaders_1.getNoCacheHeaders()));
            response.write(yield this.getCategoriesResponse());
            response.end();
        });
    }
    getCategoriesResponse() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const meta = yield this.getMetadata();
            const response = meta.categorizedComponents.map((category, index) => ({
                id: index + 1,
                isDefault: index === 0,
                name: category.name,
                sortIndex: index,
                type: exports.COMPONENT_TYPE,
            }));
            return JSON.stringify(response);
        });
    }
    getMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return getProjectMetadata_1.getProjectMetadata(this.context.uxpinDirPath);
        });
    }
}
exports.GetCategoriesHandler = GetCategoriesHandler;
