// @uxpin/merge-cli/src/steps/serialization/component/presets/jsx/compile/generateVirtualModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateVirtualModules = void 0;
const lodash_1 = require("lodash");
const path_1 = require("path");
const getComponentNamespacedName_1 = require("../../../name/getComponentNamespacedName");
const getNamespacedComponentsTree_1 = require("./getNamespacedComponentsTree");
function generateVirtualModules(components) {
    const tree = getNamespacedComponentsTree_1.getNamespacedComponentsTree(components);
    return components.map((component) => createVirtualModule(component, tree));
}
exports.generateVirtualModules = generateVirtualModules;
function createVirtualModule(component, tree) {
    const exportName = component.defaultExported ? 'default' : component.name;
    return ({
        moduleSource: `
Object.defineProperty(exports, "__esModule", { value: true });
exports.${exportName} = ${JSON.stringify(createComponentPlaceholder(component, tree))};`,
        path: removeExtensionFromPath(component.info.implementation.path),
    });
}
function createComponentPlaceholder(component, tree) {
    return lodash_1.get(tree, getComponentNamespacedName_1.getComponentNamespacedName(component));
}
function removeExtensionFromPath(path) {
    const { dir, name } = path_1.parse(path);
    return path_1.join(dir, name);
}
