// @uxpin/merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPropTypeValue = void 0;
const lodash_1 = require("lodash");
const getDefaultValueForPropType_1 = require("./getDefaultValueForPropType");
var TYPES;
(function (TYPES) {
    TYPES["dictionary"] = "dictionary";
    TYPES["literal"] = "literal";
    TYPES["shape"] = "shape";
    TYPES["typedArray"] = "typedArray";
    TYPES["union"] = "union";
})(TYPES || (TYPES = {}));
function getPropTypeValue(propertyType) {
    if (!propertyType) {
        return '';
    }
    switch (propertyType.name) {
        case TYPES.union:
            return getPropTypeValue(propertyType.structure.elements[0]);
        case TYPES.literal:
            return propertyType.structure.value;
        case TYPES.typedArray:
            return [getPropTypeValue(propertyType.structure.memberType)];
        case TYPES.dictionary:
            return getPropTypeValue(propertyType.structure.valueType);
        case TYPES.shape:
            return lodash_1.reduce(propertyType.structure, (result, value, propName) => (Object.assign(result, { [propName]: getPropTypeValue(value) })), {});
        default:
            return getDefaultValueForPropType_1.getDefaultValueForPropType(propertyType.name);
    }
}
exports.getPropTypeValue = getPropTypeValue;
