// @uxpin/merge-cli/src/steps/serialization/component/implementation/ComponentPropertyDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomControlTypeName = exports.CustomDescriptorsTags = void 0;
var CustomDescriptorsTags;
(function (CustomDescriptorsTags) {
    CustomDescriptorsTags["BIND"] = "@uxpinbind";
    CustomDescriptorsTags["DESCRIPTION"] = "@uxpindescription";
    CustomDescriptorsTags["HIDDEN"] = "@uxpinignoreprop";
    CustomDescriptorsTags["NAME"] = "@uxpinpropname";
    CustomDescriptorsTags["TYPE"] = "@uxpincontroltype";
})(CustomDescriptorsTags = exports.CustomDescriptorsTags || (exports.CustomDescriptorsTags = {}));
var CustomControlTypeName;
(function (CustomControlTypeName) {
    CustomControlTypeName["CodeEditor"] = "codeeditor";
    CustomControlTypeName["Input"] = "input";
    CustomControlTypeName["Interactions"] = "interactions";
    CustomControlTypeName["Number"] = "number";
    CustomControlTypeName["Select"] = "select";
    CustomControlTypeName["Switcher"] = "switcher";
    CustomControlTypeName["Textfield"] = "textfield";
})(CustomControlTypeName = exports.CustomControlTypeName || (exports.CustomControlTypeName = {}));
