// @uxpin/merge-cli/src/steps/serialization/component/props/ParsedPropertyModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParsedPropertyModel = void 0;
const tslib_1 = require("tslib");
const ParsedPropertyDescriptor_1 = require("../implementation/ParsedPropertyDescriptor");
class ParsedPropertyModel {
    constructor(_a) {
        var { warnings } = _a, _b = _a.result, { descriptors } = _b, definition = tslib_1.__rest(_b, ["descriptors"]);
        this.warnings = warnings;
        this.descriptors = descriptors;
        this.propertyDefinition = definition;
    }
    get name() {
        return this.propertyDefinition.name;
    }
    hasAutoUpdate() {
        return !!this.autoUpdate;
    }
    setAutoUpdate(autoUpdate) {
        this.autoUpdate = autoUpdate;
    }
    getBindingDescriptor() {
        return this.descriptors.find(ParsedPropertyDescriptor_1.isBindingDescriptor);
    }
    serialize() {
        return {
            result: Object.assign(Object.assign(Object.assign({}, this.propertyDefinition), this.serializeDescriptors()), this.serializeAutoUpdate()),
            warnings: this.warnings,
        };
    }
    serializeDescriptors() {
        return this.descriptors.reduce((result, descriptor) => {
            Object.assign(result, descriptor.serialized);
            return result;
        }, {});
    }
    serializeAutoUpdate() {
        if (this.autoUpdate) {
            return { autoUpdate: this.autoUpdate };
        }
        return {};
    }
}
exports.ParsedPropertyModel = ParsedPropertyModel;
