// @uxpin/merge-cli/src/steps/serialization/component/implementation/javascript/parsePropertyItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePropertyItem = void 0;
const pReduce = require("p-reduce");
const getDefaultValue_1 = require("./defaultValue/getDefaultValue");
const getPropertyCustomDescriptorsWithWarnings_1 = require("./props/getPropertyCustomDescriptorsWithWarnings");
const getPropertyDescriptionWithWarnings_1 = require("./props/getPropertyDescriptionWithWarnings");
const getPropertyTypeWithWarnings_1 = require("./type/getPropertyTypeWithWarnings");
function parsePropertyItem(propName, propItem) {
    const partialProviders = [
        getDefaultValue_1.getDefaultValue(propName, propItem),
        getPropertyTypeWithWarnings_1.getPropertyTypeWithWarnings(propName, propItem),
        getPropertyCustomDescriptorsWithWarnings_1.getPropertyCustomDescriptorsWithWarnings(propItem),
        getPropertyDescriptionWithWarnings_1.getPropertyDescriptionWithWarnings(propName, propItem),
    ];
    const aggregator = {
        result: {
            description: propItem.description,
            descriptors: [],
            isRequired: propItem.required,
            name: propName,
        },
        warnings: [],
    };
    return pReduce(partialProviders, (result, partial) => {
        Object.assign(result.result, partial.result);
        Object.assign(result.warnings, partial.warnings);
        return result;
    }, aggregator);
}
exports.parsePropertyItem = parsePropertyItem;
