// @uxpin/merge-cli/src/steps/serialization/component/implementation/javascript/serializeJSComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeJSComponent = void 0;
const tslib_1 = require("tslib");
const lodash_1 = require("lodash");
const joinWarningLists_1 = require("../../../../../common/warning/joinWarningLists");
const validateWrappers_1 = require("../../../validation/validateWrappers");
const getCommentTag_1 = require("../../comments/getCommentTag");
const getJSDocTagsArrayFromString_1 = require("../../comments/getJSDocTagsArrayFromString");
const CommentTags_1 = require("../../CommentTags");
const serializeAndValidateParsedProperties_1 = require("../../props/serializeAndValidateParsedProperties");
const parseWrapperAnnotation_1 = require("../../wrappers/parseWrapperAnnotation");
const getComponentDocUrlFromDescription_1 = require("./getComponentDocUrlFromDescription");
const getComponentName_1 = require("./getComponentName");
const getComponentNamespaceFromDescription_1 = require("./getComponentNamespaceFromDescription");
const getDefaultComponentFrom_1 = require("./getDefaultComponentFrom");
const isDefaultExported_1 = require("./isDefaultExported");
const parsePropertyItem_1 = require("./parsePropertyItem");
function serializeJSComponent(component) {
    return getDefaultComponentFrom_1.getDefaultComponentFrom(component.path)
        .then(thunkGetMetadata(component))
        .then(thunkGetSummaryResult(component.path));
}
exports.serializeJSComponent = serializeJSComponent;
function thunkGetMetadata(implInfo) {
    return (parsed) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        const properties = yield getComponentProperties(parsed.props);
        const name = getComponentName_1.getComponentName(implInfo.path, parsed);
        const wrappers = getComponentWrappers(parsed, implInfo);
        const defaultExported = isDefaultExported_1.isDefaultExported(implInfo.path, name);
        return Object.assign({ defaultExported,
            name,
            properties,
            wrappers }, getValuesFromComments(name, parsed));
    });
}
function getComponentProperties(props) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const propertyParsingPromises = lodash_1.toPairs(props)
            .map(([propName, propType]) => parsePropertyItem_1.parsePropertyItem(propName, propType));
        const parsedProps = yield Promise.all(propertyParsingPromises);
        return serializeAndValidateParsedProperties_1.serializeAndValidateParsedProperties(parsedProps);
    });
}
function getComponentWrappers(parsed, implInfo) {
    const jsDocTags = getJSDocTagsArrayFromString_1.getJSDocTagsArrayFromString(parsed.description);
    const wrappersTag = getCommentTag_1.getCommentTag(CommentTags_1.CommentTags.UXPIN_WRAPPERS, jsDocTags) || '';
    const wrappers = parseWrapperAnnotation_1.parseWrapperAnnotation(wrappersTag);
    return validateWrappers_1.validateWrappers(wrappers, implInfo);
}
function getValuesFromComments(name, parsed) {
    const jsDocTags = getJSDocTagsArrayFromString_1.getJSDocTagsArrayFromString(parsed.description);
    const namespaceTag = getCommentTag_1.getCommentTag(CommentTags_1.CommentTags.UXPIN_NAMESPACE, jsDocTags) || '';
    const componentDocUrlTag = getCommentTag_1.getCommentTag(CommentTags_1.CommentTags.UXPIN_DOC_URL, jsDocTags) || '';
    const namespace = getComponentNamespaceFromDescription_1.getComponentNamespaceFromDescription(name, namespaceTag);
    const componentDocUrl = getComponentDocUrlFromDescription_1.getComponentDocUrlFromDescription(componentDocUrlTag);
    return { namespace, componentDocUrl };
}
function thunkGetSummaryResult(path) {
    return (_a) => {
        var { properties, wrappers } = _a, directProps = tslib_1.__rest(_a, ["properties", "wrappers"]);
        return ({
            result: Object.assign(Object.assign({}, directProps), { properties: properties.map((p) => p.result), wrappers: wrappers.result }),
            warnings: joinWarningLists_1.joinWarningLists([
                ...properties.map((p) => p.warnings),
                wrappers.warnings,
            ], path),
        });
    };
}
