// @uxpin/merge-cli/src/program/command/push/getBuildOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBuildOptions = void 0;
const isDevelopmentEnv_1 = require("../../../program/env/isDevelopmentEnv");
const isTestEnv_1 = require("../../../program/env/isTestEnv");
const getProjectRoot_1 = require("../../args/providers/paths/getProjectRoot");
const getTempDirPath_1 = require("../../args/providers/paths/getTempDirPath");
const TEST_UXPIN_API_DOMAIN = '0.0.0.0';
function getDefaultApiDomain(domain) {
    return isTestEnv_1.isTestEnv() || isDevelopmentEnv_1.isDevelopmentEnv()
        ? process.env.UXPIN_API_DOMAIN || TEST_UXPIN_API_DOMAIN
        : `api.${domain}`;
}
function getBuildOptions(args) {
    const { token, uxpinDomain, webpackConfig, wrapper, branch, tag } = args;
    return {
        branch,
        projectRoot: getProjectRoot_1.getProjectRoot(args),
        tag,
        token,
        uxpinApiDomain: getDefaultApiDomain(uxpinDomain),
        uxpinDirPath: getTempDirPath_1.getTempDirPath(args),
        uxpinDomain,
        webpackConfigPath: webpackConfig,
        wrapperPath: wrapper,
    };
}
exports.getBuildOptions = getBuildOptions;
