// @uxpin/merge-cli/src/common/services/UXPin/createTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTag = void 0;
const tslib_1 = require("tslib");
const isTestEnv_1 = require("../../../program/env/isTestEnv");
const requestPromiseWithEnhancedError_1 = require("../../../utils/requestPromiseWithEnhancedError");
const getAuthHeaders_1 = require("./headers/getAuthHeaders");
const getUserAgentHeaders_1 = require("./headers/getUserAgentHeaders");
function createTag(opts) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        // Skip updating repository pointers in test environment
        if (isTestEnv_1.isTestEnv()) {
            return Promise.resolve();
        }
        return requestPromiseWithEnhancedError_1.requestPromiseWithEnhancedError(`${opts.apiDomain}/code/v/1.0/create-tag`, {
            body: {
                commitHash: opts.commitHash,
                tagName: opts.tag,
            },
            headers: Object.assign(Object.assign({}, getAuthHeaders_1.getAuthHeaders(opts.authToken)), getUserAgentHeaders_1.getUserAgentHeaders()),
            json: true,
            method: 'POST',
        }).then(() => undefined);
    });
}
exports.createTag = createTag;
