// @uxpin/merge-cli/src/common/services/UXPin/postUploadBundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postUploadBundle = void 0;
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const requestPromiseWithEnhancedError_1 = require("../../../utils/requestPromiseWithEnhancedError");
const getAuthHeaders_1 = require("./headers/getAuthHeaders");
const getUserAgentHeaders_1 = require("./headers/getUserAgentHeaders");
function postUploadBundle(domain, token, commitHash, path) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return requestPromiseWithEnhancedError_1.requestPromiseWithEnhancedError(`${domain}/code/v/1.0/push/bundle`, {
            formData: {
                bundle: fs_1.createReadStream(path),
                commitHash,
            },
            headers: Object.assign(Object.assign({}, getAuthHeaders_1.getAuthHeaders(token)), getUserAgentHeaders_1.getUserAgentHeaders()),
            json: true,
            method: 'POST',
        });
    });
}
exports.postUploadBundle = postUploadBundle;
