// @uxpin/merge-cli/src/program/command/experimentation/getExperimentationWatchCommandSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExperimentationWatchCommandSteps = exports.COMPILATION_SUCCESS_MESSAGE = void 0;
const print_1 = require("../../../utils/console/print");
const printLine_1 = require("../../../utils/console/printLine");
const PrintOptions_1 = require("../../../utils/console/PrintOptions");
const experimentationBuildLibraryStep_1 = require("./steps/experimentationBuildLibraryStep");
const experimentationPrepareMetadataStep_1 = require("./steps/experimentationPrepareMetadataStep");
exports.COMPILATION_SUCCESS_MESSAGE = 'ðŸŒ Compiled successfully! Now you can refresh your browser.';
function getExperimentationWatchCommandSteps(args) {
    return [
        onCompilationStart(),
        experimentationBuildLibraryStep_1.experimentationBuildLibraryStep(args),
        experimentationPrepareMetadataStep_1.experimentationPrepareMetadataStep(args),
        onCompilationEnd(),
    ];
}
exports.getExperimentationWatchCommandSteps = getExperimentationWatchCommandSteps;
function onCompilationStart() {
    return {
        exec: () => print_1.print('ðŸŽ© Hang on to your hat! Compilation in progress...', { color: PrintOptions_1.PrintColor.YELLOW }),
        shouldRun: true,
    };
}
function onCompilationEnd() {
    return {
        exec: () => printLine_1.printLine(`\r${exports.COMPILATION_SUCCESS_MESSAGE}`),
        shouldRun: true,
    };
}
