// @uxpin/merge-cli/src/steps/serialization/component/implementation/typescript/__tests__/serializeTSComponent-Annotations.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const serializeTSComponent_1 = require("../serializeTSComponent");
const serializeTSComponent_test_1 = require("./serializeTSComponent.test");
describe('serializeTSComponent-Annotations', () => {
    describe('function with bind annotation', () => {
        it('serializes correctly including bind declaration in both function and bound property', () => {
            // given
            const component = serializeTSComponent_test_1.getImplementation('FunctionWithBindAnnotation');
            // when
            return serializeTSComponent_1.serializeTSComponent(component).then((serializedProps) => {
                // then
                const expectedMetadata = {
                    defaultExported: false,
                    name: 'FunctionWithBindAnnotation',
                    properties: [
                        {
                            defaultValue: { value: false },
                            description: '',
                            isAutoUpdated: true,
                            isRequired: false,
                            name: 'isChecked',
                            type: { name: 'boolean', structure: {} },
                        },
                        {
                            description: '',
                            isRequired: true,
                            name: 'label',
                            type: { name: 'string', structure: {} },
                        },
                        {
                            defaultValue: { value: '' },
                            description: '',
                            isRequired: false,
                            name: 'name',
                            type: { name: 'string', structure: {} },
                        },
                        {
                            autoUpdate: {
                                targetPropName: 'isChecked',
                                valuePath: '0.target.checked',
                            },
                            description: '',
                            isRequired: false,
                            name: 'onChange',
                            type: { name: 'func', structure: {} },
                        },
                    ],
                    wrappers: [],
                };
                expect(serializedProps.result).toEqual(expectedMetadata);
                expect(serializedProps.warnings).toEqual([]);
            });
        });
    });
});
