// @uxpin/merge-cli/src/steps/serialization/component/implementation/typescript/__tests__/serializeTSComponent-DocUrl.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const serializeTSComponent_1 = require("../serializeTSComponent");
const serializeTSComponent_test_1 = require("./serializeTSComponent.test");
describe('serializeTSComponent-DocUrl', () => {
    describe('functional component', () => {
        it('is serialized correctly with the provided url', () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
            // having
            const component = serializeTSComponent_test_1.getImplementation('FunctionWithDocUrlDeclaration');
            const expectedMetadata = {
                componentDocUrl: 'https://app.uxpin.com/test',
                defaultExported: true,
                name: 'FunctionWithDocUrlDeclaration',
                properties: [
                    {
                        description: '',
                        isRequired: true,
                        name: 'name',
                        type: { name: 'string', structure: {} },
                    },
                ],
                wrappers: [],
            };
            // when
            const metadata = yield serializeTSComponent_1.serializeTSComponent(component);
            // then
            expect(metadata.warnings).toEqual([]);
            expect(metadata.result).toEqual(expectedMetadata);
        }));
    });
});
